src/compiler/emitter.ts

```diff
@@ -3470,6 +3494,148 @@ module ts {
+                // _i = 0
+                emitNode(counter);
+                write(" = 0");
+                emitEnd(node.expression);
+                
+                if (!rhsIsIdentifier) {
+                    // , _a = expr
+                    write(", ");
+                    emitStart(node.expression);
+                    emitNode(rhsReference);
+                    write(" = ");
+                    emitNode(node.expression);
+                    emitEnd(node.expression);
+                }
+                write("; ");
+                
+                // _i < _a.length;
+                emitStart(node.initializer);
+                emitNode(counter);
+                write(" < ");
+                emitNode(rhsReference);
+                write(".length");
+                emitEnd(node.initializer);
+                write("; ");
+                
+                // _i++)
+                emitStart(node.initializer);
+                emitNode(counter);
+                write("++");
+                emitEnd(node.initializer);
+                emitToken(SyntaxKind.CloseParenToken, node.expression.end);
+                
+                // Body
+                write(" {");
+                writeLine();
+                increaseIndent();
+                
+                // Initialize LHS
+                // var v = _a[_i];
+                var rhsIterationValue = createElementAccessExpression(rhsReference, counter);
+                emitStart(node.initializer);
+                if (node.initializer.kind === SyntaxKind.VariableDeclarationList) {
+                    write("var ");
+                    var variableDeclarationList = <VariableDeclarationList>node.initializer;
+                    if (variableDeclarationList.declarations.length >= 1) {
+                        var declaration = variableDeclarationList.declarations[0];
+                        if (isBindingPattern(declaration.name)) {
+                            // This works whether the declaration is a var, let, or const.
+                            // It will use rhsIterationValue _a[_i] as the initializer.
+                            emitDestructuring(declaration, rhsIterationValue);
+                        }
+                        else {
+                            // The following call does not include the initializer, so we have
+                            // to emit it separately.
+                            emitNode(declaration);
+                            write(" = ");
+                            emitNode(rhsIterationValue);
+                        }
+                    }
+                    else {
+                        // It's an empty declaration list. This can only happen in an error case, if the user wrote
+                        //     for (var of []) {}
+                        var emptyDeclarationListTemp = createTempVariable(node, /*forLoopVariable*/ false);
+                        emitNode(emptyDeclarationListTemp);

```

Comment:
Why the `emptyDeclarationListTemp` variable? Just pass the result of `createTempVariable` as an arg.